### PR TITLE
server: support tag sql in pprof result

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

we are hard to find which SQL real consume CPU usage when we struggle with "CPU 100% problem".

after go1.9, pprof support labels that let use can tag some kv pair in pprof result, so maybe tag sql to pprof to help us. 

and we find it.. https://rakyll.org/profiler-labels/

### What is changed and how it works?

profile labels is not free, it has some cost to call runtime set  goroutine, so this pr add config item and combine with our  `http://127.0.0.1:10080/reload-config` to hot reload,

```
[performance]
pprof-sql-cpu = false
```
default value is `false`, just set it to `true` and `curl http://127.0.0.1:10080/reload-config` when  we want to pprof sql cpu.

then using normal `curl http://0.0.0.0:10080/debug/pprof/profile --output p` or `/debug/pprof/zip` to get profile.

then you can observe tags in pprof

```
go tool pprof  p1                                        
File: ___go_build_github_com_pingcap_tidb_tidb_server
Build ID: b1542297fdc1541f774e2ab7c6cc62de595ee7f1
Type: cpu
Time: Jan 2, 2020 at 11:55am (CST)
Duration: 30s, Total samples = 300ms (    1%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) tags
 sql: Total 40.0ms
      20.0ms (50.00%): insert into t ( a ) values ( ? )
      20.0ms (50.00%): select * from t where a > ?
```

image

and remember to set `pprof-sql-cpu ` back to `false` and `curl http://127.0.0.1:10080/reload-config` after pprof.

what doesn't work?

it only can tag sql the arrival after set `pprof-sql-cpu` to true..(this is the reason why we don't turn switch automatic in `/debug/pprof/zip`)

### Check List 

Tests 

 - Manual test 

```
set conf and run pprof
```

Code changes

- add config item
- add reload
- setup labels 

Side effects

 - n/a

Related changes

 - Need to cherry-pick to the release branch

Release note

 - support tag sql in pprof result.

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/14312)

### CODE REVIEW

File: tidb-server/main.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 	"context"
 	"flag"
 	"fmt"
+	"github.com/pingcap/tidb/util"
```
Comment (qw4990) [inline tidb-server/main.go]: Please format this.

File: server/conn.go [hunk 2]
```diff
@@ -804,6 +807,12 @@ func (cc *clientConn) dispatch(ctx context.Context, data []byte) error {
 	cc.lastPacket = data
 	cmd := data[0]
 	data = data[1:]
+	if util.EnablePProfSQLCPU.Load() {
+		defer pprof.SetGoroutineLabels(ctx)
+		lastSQL := getLastStmtInConn{cc}.String()
+		ctx = pprof.WithLabels(ctx, pprof.Labels("sql", parser.Normalize(lastSQL)))
+		pprof.SetGoroutineLabels(ctx)
```
Comment (qw4990) [inline server/conn.go]: Just a question: if this goroutine starts some new goroutines, should we call this function with the same ctx label in these new goroutines when they start?

Comment (lysu) [reply server/conn.go]: I do some Investigate, new goroutines will inherit labels set by parent goroutine https://github.com/golang/go/blob/master/src/runtime/proc.go#L3466 :open_mouth:  maybe it's ok for us

Comment (qw4990) [reply server/conn.go]: SO COOL...!

**Comment by lysu** [comment]
## Benchmark Report

> Run Sysbench Performance Test on VMs

```diff
@@                               Benchmark Diff                               @@
================================================================================
--- tidb: b398ef4d2ead6ae98287d9489976a7dc8dba8e49
+++ tidb: 91144ca41fcd86801e85ca8be2d748f3b3885568
tikv: 1211ed161c093ab180e42879146dedd05c4c6fb5
pd: 4498b6f203588f556cb6f6ebcdcb55f0afff0efa
================================================================================
oltp_update_index:
    * QPS: 4246.60 ± 0.04% (std=1.15) delta: 0.06% (p=0.363)
    * Latency p50: 30.14 ± 0.03% (std=0.01) delta: -0.07%
    * Latency p99: 54.84 ± 3.64% (std=1.22) delta: -1.81%
            
oltp_insert:
    * QPS: 7567.92 ± 0.04% (std=2.21) delta: 0.47% (p=0.029)
    * Latency p50: 16.91 ± 0.04% (std=0.00) delta: -0.49%
    * Latency p99: 29.07 ± 4.09% (std=0.78) delta: -1.76%
            
oltp_read_write:
    * QPS: 15694.90 ± 0.25% (std=27.78) delta: -0.52% (p=0.072)
    * Latency p50: 163.45 ± 0.21% (std=0.25) delta: 0.55%
    * Latency p99: 326.02 ± 1.81% (std=4.79) delta: 0.90%
            
oltp_point_select:
    * QPS: 41161.17 ± 0.17% (std=53.14) delta: -3.90% (p=0.020)
    * Latency p50: 3.11 ± 0.21% (std=0.00) delta: 4.02%
    * Latency p99: 10.09 ± 0.00% (std=0.00) delta: 3.70%
            
oltp_update_non_index:
    * QPS: 4715.66 ± 0.83% (std=28.71) delta: -0.12% (p=0.341)
    * Latency p50: 27.14 ± 0.84% (std=0.17) delta: 0.14%
    * Latency p99: 43.92 ± 2.41% (std=0.75) delta: 2.54%
            
```

_Originally posted by @sre-bot in https://github.com/pingcap/tidb/pull/14362#issuecomment-571173870_

**Pull request merged**